core/tasks.rs

```diff
@@ -0,0 +1,273 @@
+use futures::task::AtomicWaker;
+use std::marker::PhantomData;
+use std::ops::DerefMut;
+use std::sync::atomic::AtomicBool;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::task::Context;
+use std::task::Poll;
+
+type UnsendTask = Box<dyn FnOnce(&mut v8::HandleScope) + 'static>;
+type SendTask = Box<dyn FnOnce(&mut v8::HandleScope) + Send + 'static>;
+
+static_assertions::assert_not_impl_any!(V8TaskSpawnerFactory: Send);
+static_assertions::assert_not_impl_any!(V8TaskSpawner: Send);
+static_assertions::assert_impl_all!(V8CrossThreadTaskSpawner: Send);
+
+#[derive(Default)]
+pub(crate) struct V8TaskSpawnerFactory {
+  // TODO(mmastrac): ideally we wouldn't box if we could use arena allocation and a max submission size
+  // TODO(mmastrac): we may want to split the Send and !Send tasks
+  /// The set of tasks, non-empty if `has_tasks` is set.
+  tasks: Mutex<Vec<SendTask>>,
+  /// A flag we can poll without any locks.
+  has_tasks: AtomicBool,
+  /// The polled waker, woken on task submission.
+  waker: AtomicWaker,
+  /// This cannot be Send because it may contain `!Send` tasks submitted by a [`V8TaskSpawner`]. It is
+  /// only safe to send this object to another thread if you plan on submitting [`Send`] tasks to it,
+  /// which is what [`V8CrossThreadTaskSpawner`] does.
+  _unsend_marker: PhantomData<*const ()>,
+}
```

Comment:
```suggestion
/// V8TaskSpawnerFactory cannot be Send because it may contain `!Send` tasks 
/// submitted by a [`V8TaskSpawner`]. Use [`V8CrossThreadTaskSpawner`] to
/// submit [`Send`] tasks.
#[derive(Default)]
pub(crate) struct V8TaskSpawnerFactory {
  // TODO(mmastrac): ideally we wouldn't box if we could use arena allocation and a max submission size
  // TODO(mmastrac): we may want to split the Send and !Send tasks
  /// The set of tasks, non-empty if `has_tasks` is set.
  tasks: Mutex<Vec<SendTask>>,
  /// A flag we can poll without any locks.
  has_tasks: AtomicBool,
  /// The polled waker, woken on task submission.
  waker: AtomicWaker,
  _unsend_marker: PhantomData<*const ()>,
}
```

Reply:
Tweaked comment with a similar message